README.md
# gauges

![](doc/demo.png)

*Note: This project serves as a training exercise for myself in order to better understand the inner workings of Flutter and learn how to write custom `Widget`s, `Element`s and `RenderObject`s from scratch. As a result, the API might change heavily and I may not know about best practices or optimal solutions for many problems. So if you have any suggestions of improvement, please feel free to open an issue.*


## Getting started

To use this plugin, add [flutter_gauges] as a [dependency in your pubspec.yaml file](https://flutter.dev/docs/development/packages-and-plugins/using-packages).

## Usage

### Radial gauge

The basic widget for radial gauges is the `RadialGauge`. Multiple `axes` can be added. E. g. this is a gauge with three axes:

<img src="https://raw.githubusercontent.com/JulianAssmann/flutter_gauges/master/doc/multiple_axes.png" width="200" height="200">

This is the basic structure of a `RadialGauge`. See the following sections for more info on each of them.

```dart
RadialGauge(
  axes: [
    RadialGaugeAxis(
        segments: [
            RadialGaugeSegment(),
            RadialGaugeSegment(),
            // ...
        ],
        pointers: [
            RadialNeedlePointer(),
            RadialNeedlePointer(),
            // ...
        ],
        ticks: [
            RadialTicks(
                children: [
                    RadialTick(),
                    RadialTick(),
                ],
            ),
            RadialTick(),
            // ...
    ),
    RadialGaugeAxis(
    ),
  ],
)
```

#### Axes

An axes has a few important properties:

- `min`: The minimum value of this axes
- `max`: The maximum value of this axes
- `minAngle`: The angle in degrees `min` is mapped to
- `maxAngle`: The angle in degrees `max` is mapped to
- `radius`: The radius of the axes as a fraction of the `RadialGauge`'s size
- `width`: The width of the axes as a fraction of the `RadialGauge`'s size
- `offset`: The offset of the axes in respect to the middle of the `RadialGauge`
- `color`: The color of the axes (gradients are also available, see examples)

For more, take a look at the examples.
 
```dart
RadialGauge(
  axes: [
    RadialGaugeAxis(
      minValue: -100,
      maxValue: 100,
      minAngle: -150,
      maxAngle: 150,
      radius: 0.6,
      width: 0.2,
    ),
  ],
)
```

#### Segments

Each axes can consist of multiple segments (it is, in fact, a segment itself). Each of the segments has a properties similiar to the `RadialGaugeAxis`. The segments are listed in the `segments` property of an axes:

```dart
RadialGauge(
  axes: [
    RadialGaugeAxis(
      color: Colors.transparent,
      // ...
      segments: [
        RadialGaugeSegment(
          minValue: 0, 
          maxValue: 20, 
          minAngle: -150, 
          maxAngle: -130,
          color: Colors.red,
        ),
        RadialGaugeSegment(
          minValue: 20, 
          maxValue: 40, 
          minAngle: -120, 
          maxAngle: -90,
          color: Colors.orange,
        ),
          // ...
      ],
    ),
  ],
)
```

<img src="https://raw.githubusercontent.com/JulianAssmann/flutter_gauges/master/doc/segments.png" width="200" height="200">

#### Ticks

Each axes can have multiple ticks of type `RadialTick`. They are listed in the `ticks` property. 
There are multiple possibilities to create ticks.

- Interval: Set the interval to the desired value. The ticks are displayed on every integer multiple of `interval`.
- Ticks in between: Set `ticksInBetween` to the desired number of ticks in between the parent ticks or the axes limits.
- Custom values: Set `values` to the desired values you want the ticks to appear at.
- Custom builder: Set `valuesBuilder` and/or `anglesBuilder` to a custom function calculating the values and/or angles at which the ticks should be displayed.

Every `RadialTick` can have `RadialTicks` as children, which is especially useful when setting the `ticksInBetween` on a children tick.

```dart
RadialGaugeAxis(
  // ...
  color: Colors.transparent,
  ticks: [
    RadialTicks(
        interval: 20,
        alignment: RadialTickAxisAlignment.inside,
        color: Colors.blue,
        length: 0.2,
        children: [
          RadialTicks(
              ticksInBetween: 5,
              length: 0.1,
              color: Colors.grey[500]),
        ])
  ],
)
```

<img src="https://raw.githubusercontent.com/JulianAssmann/flutter_gauges/master/doc/ticks_in_between.png" width="200" height="200">

#### Pointer

Each axis can have multiple pointers. Currently, there is only one type of pointer available, the `RadialNeedlePointer` with the following properties:

- `value`: The value the pointer is pointing to
- `ticknessStart`: The thickness of the pointer at the start of the pointer (in the middle of the axis)
- `ticknessEnd`: The thickness of the pointer at the end of the pointer
- `length`: The length of the pointer in fractions of the axis.
- `knobRadiusAboslute`: The radius of the pointer's knob in absolute pixels.
- `color`/`gradient`: The color/gradient filling the pointer.

The following code results in the knob shown above. For more, take a look at the examples.

```dart
RadialGaugeAxis(
    pointers: [
        RadialNeedlePointer(
            value: _pointerValue,
            thicknessStart: 20,
            thicknessEnd: 0,
            length: 0.6,
            knobRadiusAbsolute: 10,
            gradient: LinearGradient(...),
        )
    ],
)
```

### Linear gauge

Not available yet

## Maintainer
[Julian Aßmann](https://github.com/JulianAssmann)



CHANGELOG.md
## [0.0.1+1] - 21.06.2021

* Migrate to null safety

## [0.0.1+1] - 18.06.2021

* Fix README to correctly display images

## [0.0.1] - 18.06.2021

* Initial release (including basic radial gauge functionality)


example/lib/pages/multiple_axes_page.dart
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:gauges/gauges.dart';
import 'package:url_launcher/url_launcher.dart';

class MultipleAxesPage extends StatefulWidget {
  const MultipleAxesPage({Key? key}) : super(key: key);

  @override
  _MultipleAxesPageState createState() => _MultipleAxesPageState();
}

class _MultipleAxesPageState extends State<MultipleAxesPage> {
  double _pointerValue = 0;

  @override
  Widget build(BuildContext context) {
    final size = min(MediaQuery.of(context).size.width,
            MediaQuery.of(context).size.height) -
        100;
    return Scaffold(
      appBar: AppBar(
        title: Text('Multiple axes'),
        actions: [
          IconButton(
            icon: Icon(Icons.code_outlined),
            onPressed: () async {
              await launch(
                  'https://github.com/JulianAssmann/flutter_gauges/tree/master/example/lib/pages/multiple_axes_page.dart');
            },
          ),
        ],
      ),
      body: Column(
        children: [
          SizedBox(
            width: size,
            height: size,
            child: Padding(
              padding: const EdgeInsets.all(20.0),
              child: RadialGauge(
                axes: [
                  // Left axis
                  RadialGaugeAxis(
                    minValue: -100,
                    maxValue: 0,
                    minAngle: -150,
                    maxAngle: 0,
                    radius: 0.15,
                    width: 0.05,
                    offset: Offset(-0.2, -0.1),
                    color: Colors.red[300],
                    pointers: [
                      RadialNeedlePointer(
                        value: -100 - _pointerValue,
                        thickness: 4.0,
                        knobColor: Colors.transparent,
                        length: 0.2,
                      ),
                    ],
                    ticks: [
                      RadialTicks(
                          alignment: RadialTickAxisAlignment.below,
                          ticksInBetween: 10,
                          length: 0.05)
                    ],
                  ),
                  // Left axis
                  RadialGaugeAxis(
                    minValue: -100,
                    maxValue: 0,
                    minAngle: 0,
                    maxAngle: 90,
                    radius: 0.15,
                    width: 0.05,
                    offset: Offset(0.2, -0.1),
                    color: Colors.green[300],
                    pointers: [
                      RadialNeedlePointer(
                        value: -100 - _pointerValue,
                        thickness: 4.0,
                        knobColor: Colors.transparent,
                        length: 0.2,
                      ),
                    ],
                    ticks: [
                      RadialTicks(
                          alignment: RadialTickAxisAlignment.above,
                          ticksInBetween: 10,
                          length: 0.05)
                    ],
                  ),
                  // Main axis
                  RadialGaugeAxis(
                    minValue: -100,
                    maxValue: 0,
                    minAngle: -150,
                    maxAngle: 150,
                    radius: 0.6,
                    width: 0.2,
                    color: Colors.lightBlue[200],
                    ticks: [
                      RadialTicks(
                          interval: 50,
                          alignment: RadialTickAxisAlignment.inside,
                          color: Colors.black,
                          length: 0.2,
                          children: [
                            RadialTicks(
                              ticksInBetween: 5,
                              length: 0.1,
                              color: Colors.blueGrey,
                            ),
                          ])
                    ],
                    pointers: [
                      RadialNeedlePointer(
                        value: _pointerValue,
                        thicknessStart: 20,
                        thicknessEnd: 0,
                        color: Colors.lightBlue[200]!,
                        length: 0.6,
                        knobRadiusAbsolute: 10,
                      ),
                    ],
                  ),
                ],
              ),
            ),
          ),
          SizedBox(height: 24),
          Slider(
              value: _pointerValue,
              min: -100,
              max: 0,
              onChanged: (value) {
                setState(() {
                  _pointerValue = value;
                });
              }),
          SizedBox(height: 12),
          Text(_pointerValue.round().toString())
        ],
      ),
    );
  }
}
